docs/getting_started/module.md

```diff
@@ -0,0 +1,260 @@
+12:53:50.455 kernel  info    stopped background module metadata-server           application: application
+12:53:50.455 kernel  info    stopped stage 1024                                  application: application
+12:53:50.455 kernel  info    stopping stage 0                                    application: application
+12:53:50.455 kernel  info    stopped stage 0                                     application: application
+12:53:50.455 kernel  info    leaving kernel                                      application: application
+```
+
+When running the code, _"foreground-module"_ gets to complete successfully, then as the kernel has no other foreground modules, _"hello-world"_ will have its `ctx` cancelled, and thus log `Time to stop`.
+
+## Additional features
+
+### Default values
+At runtime, every _Module_ object will have to implement the following interfaces:
+
+[embedmd]:# (../../pkg/kernel/module.go /type Module interface/ /\n}/)
+```go
+type Module interface {
+	// Run the module. If the provided context is canceled you have a
+	// few seconds (configurable with kernel.killTimeout) until your module
+	// is killed (via exit(1)). If you return from Run, it is assumed that
+	// your module is done executing and (depending on the type of your module)
+	// this might trigger a kernel shutdown. If you return an error, a kernel
+	// shutdown is also triggered.
+	Run(ctx context.Context) error
+}
+```
+
+This is the backbone of Gosoline module creation. Each _ModuleFactory_ is a function that returns one such _Module_ object.
+
+[embedmd]:# (../../pkg/kernel/module.go /type TypedModule interface/ /\n}/)
+```go
+type TypedModule interface {
+	IsEssential() bool
+	IsBackground() bool
+}
+```
+
+Internally, a kernel needs to know if a module is a background module, and if it is essential. To configure these properties you need to pass in _Options_ as parameters to `kernel.Add`. The default values, for a module without any specific options, is that this module is a non-essential foreground module. 
+
+[embedmd]:# (../../pkg/kernel/module.go /type StagedModule interface/ /\n}/)
+```go
+type StagedModule interface {
+	GetStage() int
+}
+```
+
+The stage mechanism is explained below.
+
+[embedmd]:# (../../pkg/kernel/module.go /type FullModule interface/ /\n}/)
+```go
+type FullModule interface {
+	Module
+	TypedModule
+	StagedModule
+}
+```
+
+A _FullModule_ needs to implement the other three interfaces.
+
+### Module and stages
+
+When a kernel starts, not all modules are started at the same time. They are grouped into stages, and all modules in a stage are spawned at the same time. Stage one gets spawned before stage two, stage two before stage three, and so on. It is, however, the Golang scheduler that decides when to run each spawned module, thus there are no guarantees that modules in smaller stages will get to run before modules in larger stages.
```

Comment:
To be honest I would remove the whole section about the order in which they are spawned and just say they are spawned at the same time. Because you can't design code for "most of the time it will run in that order and if it does not, my code gets stuck". We only spawn them in that order because that was the easiest way to do it